### PR TITLE
Add raw request handler

### PR DESCRIPTION
**Pull request by joao-r-reis**
In order to test some edge cases around invalid or unsupported protocol versions, it is necessary to add a request handler type that returns already encoded responses to bypass the embedded encode process that the `CqlServer` does.

### CODE REVIEW

File: client/server.go [hunk 1]
```diff
@@ -334,6 +341,23 @@ func (server *CqlServer) BindAndInit(
 	}
 }
 
+type Response struct {
```
Comment (adutra) [inline client/server.go:344]: Does this new type need to be exported? It seems it's rather an internal type.

File: client/server.go [hunk 2]
```diff
@@ -334,6 +341,23 @@ func (server *CqlServer) BindAndInit(
...
+	responseFrame *frame.Frame
+	rawResponse   []byte
+}
+
+func NewFrameResponse(frameResponse *frame.Frame) *Response {
```
Comment (adutra) [inline client/server.go:349]: Same for the constructor functions, it seems they are not meant to be called from outside this package.

File: client/server.go [hunk 3]
```diff
@@ -641,19 +682,32 @@ func (c *CqlServerConnection) invokeRequestHandlers(request *frame.Frame) {
...
+				log.Debug().Msgf("%v: request handler %v produced response: %v", c, i, rawResponse)
+				if err = c.SendRaw(rawResponse); err != nil {
+					log.Error().Err(err).Msgf("%v: send failed for frame: %v", c, rawResponse)
 				}
 				break
 			}
 		}
-		if response == nil {
+		if rawResponse == nil {
 			log.Debug().Msgf("%v: no request handler could handle the request: %v", c, request)
```
Comment (adutra) [inline client/server.go:697]: Hmm this message is now misleading, since we haven't tried yet the "normal" request handlers.

**Review by adutra — CHANGES_REQUESTED** [review]
Nice feature 👍

File: client/server.go [hunk 4]
```diff
@@ -77,6 +77,10 @@ func (ctx requestHandlerContext) GetAttribute(name string) interface{} {
 // if any, may be tried.
 type RequestHandler func(request *frame.Frame, conn *CqlServerConnection, ctx RequestHandlerContext) (response *frame.Frame)
 
+// RequestRawHandler is similar to RequestHandler but returns a raw byte slice, this can be used to return responses that the
+// embedded codecs can't encode
+type RequestRawHandler func(request *frame.Frame, conn *CqlServerConnection, ctx RequestHandlerContext) (rawResponse []byte)
```
[the remark this answers is not in the document]
Comment (joao-r-reis) [reply client/server.go]: done

File: client/server.go (hunk 1 above)
Comment (joao-r-reis) [reply client/server.go]: nope, renamed it

File: client/server.go (hunk 2 above)
Comment (joao-r-reis) [reply client/server.go]: 👍  done

File: client/server.go (hunk 3 above)
Comment (joao-r-reis) [reply client/server.go]: I removed this message and kept the one after every handler type is attempted.

File: client/server.go [hunk 5]
```diff
@@ -334,6 +341,23 @@ func (server *CqlServer) BindAndInit(
...
+	rawResponse   []byte
+}
+
+func newFrameResponse(frameResponse *frame.Frame) *response {
+	return &response{
+		responseFrame: frameResponse,
+	}
+}
+
+func NewRawResponse(rawResponse []byte) *response {
```
Comment (adutra) [inline client/server.go:355]:
```suggestion
func newRawResponse(rawResponse []byte) *response {
```

**Review by adutra — CHANGES_REQUESTED** [review]
Almost there 👍

File: client/server.go (hunk 5 above)
Comment (joao-r-reis) [reply client/server.go]: I missed this one 😅

**Review by adutra — APPROVED** [review]
